File: src/dbgsprw/core/Builder.java
package dbgsprw.core;

import java.io.File;
import java.util.ArrayList;
import java.util.Map;

/**
 * Copyright 2016 dbgsprw / dbgsprw@gmail.com
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class Builder {

    ShellCommandExecutor mShellCommandExecutor;
    private ArrayList<String> mLunchMenuList;
    private String mProjectPath;
    private String mTargetProduct = "";
    private String mTargetBuildVariant = "eng";
    private String mOutDir;
    private String mTarget;
    private String mExtraArguments = "";
    private boolean mIsVerbose;
    private String mOneShotMakefile;
    private int mJobNumber = 1;
    private Process mMakeProcess;
    private boolean mIsAOSPPath;
    private String mProductOutPath;
    private MakeSetReceiver mMakeSetReceiver;

    public final static int FOUND_AOSP_HOME = 0;


    public Builder(String projectPath, MakeSetReceiver makeSetReceiver) {
        mLunchMenuList = new ArrayList<>();
        mProjectPath = projectPath;
        mShellCommandExecutor = new ShellCommandExecutor();
        mShellCommandExecutor.directory(new File(mProjectPath));
        mMakeSetReceiver = makeSetReceiver;

        updateLunchMenu();
    }

    public void setAndroidJavaHome(String directoryPath) {
        Map<String, String> env = mShellCommandExecutor.environment();
        env.put("ANDROID_JAVA_HOME", directoryPath);
        final String path = env.get("PATH");
        final String jdkBinPath = Utils.pathJoin(directoryPath, "bin");
        if (!path.startsWith(jdkBinPath)) {
            env.put("PATH", jdkBinPath + File.pathSeparator + path);
        }
    }

    public void changeProjectPath(String projectPath) {
        mProjectPath = projectPath;
        mShellCommandExecutor.directory(new File(mProjectPath));
        updateLunchMenu();
    }

    public ArrayList<String> buildMakeCommand() {
        final ArrayList<String> makeCommandLine;

        makeCommandLine = new ArrayList<>();
        makeCommandLine.add("make");

        if (mJobNumber > 1) {
            makeCommandLine.add("-j" + mJobNumber);
        }

        if (mTarget != null) {
            makeCommandLine.add(mTarget);
        }
        if (mOneShotMakefile != null) {
            makeCommandLine.add("ONE_SHOT_MAKEFILE=" + mOneShotMakefile);
        }
        if (mTargetProduct != null) {
            makeCommandLine.add("TARGET_PRODUCT=" + mTargetProduct);
        }
        if (mTargetBuildVariant != null) {
            makeCommandLine.add("TARGET_BUILD_VARIANT=" + mTargetBuildVariant);
        }

        if (mIsVerbose) {
            makeCommandLine.add("showcommands");
        }
        if (!mExtraArguments.equals("")) {
            for (String arg: mExtraArguments.split(("\\s+"))) {
                makeCommandLine.add(arg);
            }
        }
        return makeCommandLine;
    }

    public void executeMake(ShellCommandExecutor.ResultReceiver receiver) {
        mMakeProcess = mShellCommandExecutor.executeShellCommand(buildMakeCommand(), receiver);
    }

    public void stopMake() {
        if (mMakeProcess != null) {
            mMakeProcess.destroy();
            mMakeProcess = null;

        }
    }

    public ArrayList<String> getLunchMenuList() {
        return mLunchMenuList;
    }

    public void setOneShotMakefile(String directory) {
        mOneShotMakefile = Utils.pathJoin(directory, Utils.ANDROID_MK);
        mTarget = "all_modules";
    }

    public void setTarget(String target) {
        mOneShotMakefile = null;
        mTarget = target;
    }

    public void setTargetBuildVariant(String targetBuildVariant) {
        mTargetBuildVariant = targetBuildVariant;
        updateOutDir();
    }

    public void setTargetProduct(String targetProduct) {
        mTargetProduct = targetProduct;
        updateOutDir();
    }

    public void setJobNumber(int jobNumber) {
        mJobNumber = jobNumber;
    }

    private void updateOutDir() {
        mOutDir = Utils.join('-', "out", mTargetProduct, mTargetBuildVariant);
        mShellCommandExecutor.environment().put("OUT_DIR", mOutDir);
    }

    public String getOutDir() {
        return mOutDir;
    }

    public void setExtraArguments(String args) {
        mExtraArguments = args.trim();
    }

    public void setVerbose(boolean isVerbose) {
        mIsVerbose = isVerbose;
    }

    private void updateLunchMenu() {
        mLunchMenuList.clear();
        ArrayList<String> lunchCommand = new ArrayList<>();
        lunchCommand.add("source build/envsetup.sh > /dev/null ;" +
                "printf '%s\\n' ${LUNCH_MENU_CHOICES[@]} | cut -f 1 -d - | sort -u");
        mShellCommandExecutor.executeInBash(lunchCommand, new ShellCommandExecutor.ResultReceiver() {
            @Override
            public void newOut(String line) {
                if ("".equals(line)) {
                    mIsAOSPPath = false;
                } else {
                    mIsAOSPPath = true;
                    String[] lunchMenus = line.split(" ");
                    for (String lunchMenu : lunchMenus) mLunchMenuList.add(lunchMenu);
                }
            }

            @Override
            public void newError(String line) {

            }

            @Override
            public void onExit(int code) {
                mMakeSetReceiver.optionChanged(FOUND_AOSP_HOME);
            }
        });
    }

    public boolean isAOSPPath() {
        return mIsAOSPPath;
    }

    public void findOriginalProductOutPath(ShellCommandExecutor.ResultReceiver receiver) {
        String selectedTarget = mTargetProduct + '-' + mTargetBuildVariant;
        ArrayList<String> command = new ArrayList<>();
        command.add("source build/envsetup.sh > /dev/null;" +
                " lunch " + selectedTarget + " > /dev/null; echo $ANDROID_PRODUCT_OUT");
        mShellCommandExecutor.executeInBash(command, receiver);
    }

    public interface MakeSetReceiver {
        public void optionChanged(int state);
    }
}

File: src/main/java/dbgsprw/core/Builder.kt
/*
 * Copyright 2016 dbgsprw / dbgsprw@gmail.com
 * Copyright 2016 Young Ho Cha / ganadist@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package dbgsprw.core

import com.intellij.util.io.SafeFileOutputStream
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.*

/**
 * Created by ganadist on 16. 2. 25.
 */
class Builder : CommandExecutor() {
    private var mTargetProduct = ""
    private var mBuildVariant = ""
    private var mOutDir = ""
    private var mTarget = ""
    private var mTargetSdk = false
    private var mOneShotMakefile: String? = null
    private val CM_PRODUCT_PREFIX = "cm_"
    private var mLunchProcess: Process? = null

    init {
        setenv("USE_CCACHE", "1")
    }

    fun setAndroidJavaHome(home: String) {
        setenv("ANDROID_JAVA_HOME", home)
        val path = getenv("PATH")
        val jdkBinPath = arrayOf(home, "bin").joinToString(File.separator)
        if (!path.startsWith(jdkBinPath)) {
            setenv("PATH", jdkBinPath + File.pathSeparator + path)
        }
    }

    fun setTargetProduct(product: String) {
        mTargetProduct = product
        updateOutDir()
    }

    fun setBuildVariant(variant: String) {
        mBuildVariant = variant
        updateOutDir()
    }

    fun setTarget(target: String) {
        mTarget = target
        mOneShotMakefile = null
        mTargetSdk = mTarget.contains("sdk")
        updateOutDir()
    }

    fun setOneShotDirectory(directory: String) {
        mTarget = "all_modules"
        mOneShotMakefile = directory + File.separator + Utils.ANDROID_MK
    }

    private fun updateOutDir() {
        if (mTargetProduct.isNullOrBlank() || mBuildVariant.isNullOrBlank()) {
            return
        }
        var outDir = arrayOf("out", mTargetProduct, mBuildVariant).joinToString("-")
        if (mTargetSdk) {
            outDir = arrayOf(outDir, "sdk").joinToString("-")
        }
        if (outDir != mOutDir) {
            mOutPathListener.onOutDirChanged(outDir)
            mOutDir = outDir
            setenv("OUT_DIR", outDir)
            generateBuildSpec()
            if (mLunchProcess != null) {
                mLunchProcess!!.destroy()
            }
            mLunchProcess = findProductOutPath()
        }
    }

    private fun generateBuildSpec() {
        val FIRST_LINE = "# generated from AndroidBuilder\n"
        val sb = StringBuilder(FIRST_LINE)
        sb.append("TARGET_PRODUCT?=$mTargetProduct\n")
        sb.append("TARGET_BUILD_VARIANT?=$mBuildVariant\n")
        sb.append("OUT_DIR?=$mOutDir\n")
        if (mTargetProduct.startsWith(CM_PRODUCT_PREFIX)) {
            val cmBuild = mTargetProduct.substring(CM_PRODUCT_PREFIX.length)
            sb.append("CM_BUILD?=$cmBuild\n")
        }

        val specFileName = "buildspec.mk.AndroidBuilder"
        val buildSpec = File(directory(), "buildspec.mk")
        if (!buildSpec.exists() && buildSpec.createNewFile()) {
            val bos = BufferedOutputStream(SafeFileOutputStream(buildSpec)).bufferedWriter()
            bos.write(FIRST_LINE)
            bos.newLine()
            bos.write("# If you don't want to associate AndroidBuilder anymore,\n")
            bos.write("# delete following line.\n")
            bos.write("-include $specFileName\n")
            bos.close()
        }

        val buildSpecForBuilder = File(directory(), specFileName)
        buildSpecForBuilder.delete()
        if (buildSpecForBuilder.createNewFile()) {
            val bos = BufferedOutputStream(SafeFileOutputStream(buildSpecForBuilder)).bufferedWriter()
            bos.write(sb.toString())
            bos.close()
        }
    }

    fun buildMakeCommand(jobs: Int, verbose: Boolean, extras: String?): List<String> {
        val command: MutableList<String> = mutableListOf()
        command.add("make")
        if (jobs > 1) {
            command.add("-j$jobs")
        }
        command.add("TARGET_PRODUCT=$mTargetProduct")
        if (mTargetProduct.startsWith(CM_PRODUCT_PREFIX)) {
            val cmBuild = mTargetProduct.substring(CM_PRODUCT_PREFIX.length)
            command.add("CM_BUILD=$cmBuild")
            command.add("BUILD_WITH_COLORS=0") // turn off color
            command.add("CLANG_CONFIG_EXTRA_CFLAGS=-fno-color-diagnostics")
            command.add("CLANG_CONFIG_EXTRA_CPPFLAGS=-fno-color-diagnostics")
        }
        command.add("TARGET_BUILD_VARIANT=$mBuildVariant");
        if (!mOneShotMakefile.isNullOrEmpty()) {
            command.add("ONE_SHOT_MAKEFILE=$mOneShotMakefile")
        }
        command.add(mTarget)
        if (verbose) {
            command.add("showcommands")
        }
        if (!extras.isNullOrBlank()) {
            command.addAll(extras!!.split("\\s+"))
        }

        return command
    }

    fun runCombo(listener: ComboMenuListener): Process {

        val command = arrayOf("source build/envsetup.sh > /dev/null",
                "printf '%s\n' \${LUNCH_MENU_CHOICES[@]} | cut -f 1 -d - | sort -u").asList()

        return run(command, object : CommandHandler {
            override fun onOut(line: String) {
                listener.onTargetAdded(line)
            }

            override fun onExit(code: Int) {
                listener.onCompleted()
            }
        }, true)
    }

    private fun findProductOutPath(): Process {
        val selectedTarget = mTargetProduct + '-' + mBuildVariant
        Utils.log("builder", "target = $selectedTarget out_dir = $mOutDir")
        val command = arrayOf("source build/envsetup.sh > /dev/null",
                "lunch $selectedTarget > /dev/null",
                "echo \$ANDROID_PRODUCT_OUT").asList()

        return run(command, object : CommandHandler {
            override fun onOut(line: String) {
                var path = line
                Utils.log("builder", "ANDROID_PRODUCT_OUT set " + path)
                setenv("ANDROID_PRODUCT_OUT", path);
                if (!path.startsWith(File.separator)) {
                    path = directory() + File.separator + path
                }
                mOutPathListener.onAndroidProductOutChanged(path)
            }
        }, true)
    }

    private var mOutPathListener: OutPathListener = object : OutPathListener {}

    fun setOutPathListener(listener: OutPathListener) {
        mOutPathListener = listener
    }

    interface OutPathListener {
        fun onOutDirChanged(path: String) {
        }

        fun onAndroidProductOutChanged(path: String) {
        }
    }

    interface ComboMenuListener {
        fun onTargetAdded(target: String)
        fun onCompleted()
    }
}
